ambari-server/src/main/java/org/apache/ambari/server/state/ServiceGroupImpl.java

```diff
@@ -62,24 +65,32 @@
   @AssistedInject
   public ServiceGroupImpl(@Assisted Cluster cluster,
                           @Assisted("serviceGroupName") String serviceGroupName,
+                          @Assisted("version") String version,
                           @Assisted("serviceGroupDependencies") Set<ServiceGroupKey> serviceGroupDependencies,
                           ClusterDAO clusterDAO,
+                          StackDAO stackDAO,
                           ServiceGroupDAO serviceGroupDAO,
                           AmbariEventPublisher eventPublisher,
                           Clusters clusters) throws AmbariException {
 
     this.cluster = cluster;
     this.clusters = clusters;
     this.clusterDAO = clusterDAO;
+    this.stackDAO = stackDAO;
     this.serviceGroupDAO = serviceGroupDAO;
     this.eventPublisher = eventPublisher;
 
     this.serviceGroupName = serviceGroupName;
+    // version is actually stackName-stackVersion
+    int idx = version.indexOf('-');
+    StackId stackId = new StackId(version.substring(0, idx), version.substring(idx+1));
```

Comment:
I think there's still a little bit of disconnect - I think it would be clearer to separate out the mpack name and the mpack version into separate fields. That way, we don't need to parse dashes and things like that.

Reply:
Also, you could just use the ctor of StackId that does the parsing for you